scipy/optimize/tests/test_linprog.py

```diff
@@ -11,6 +11,11 @@
 from scipy._lib._numpy_compat import _assert_warns, suppress_warnings
 from scipy.sparse.linalg import MatrixRankWarning
 from scipy.linalg import LinAlgWarning
+try:
+    from scikits.umfpack import UmfpackWarning
+    has_umfpack = True
+except ImportError:
+    has_umfpack = False
```

Comment:
@mdhaber testing with `sksparse` seems to be missing here, or am I overlooking it?

Reply:
`linprog` automatically tries the solvers from CHOLMOD, UMFPACK, and SuperLU in that order. If one is not installed or otherwise fails, it moves on to the next.

So if scikit-sparse is installed, all sparse tests will start with CHOLMOD.

The check for UMFPACK here isn't to enable any UMFPACK tests; it is just to filter warnings UMFPACK generates (whenever scikit-umfpack is installed but scikit-sparse isn't).

It would be better, though, to run all the tests separately with all the solvers that are available. I was going to create subclasses that set the options accordingly, but when the scikits aren't installed this would effectively repeat the tests three times using SuperLU. Thoughts on the best way to implement this? (Can we run an entire class of tests only if an import succeeds?)

Reply:
> Can we skip a class if an import fails?

yes:
```
if has_sksparse:
    class TestLinprogSksparse(TestLinprog):
        method = 'sksparse'  # or whatever you need to force the solver
```
that should be all. so set the baseline for no sksparse and no umfpack, then subclass the main class like above (2x). and check separately that the automatic selection is working.

there are probably 10 other ways of doing the same thing ....

Reply:
OK, I was hoping that would work. (Figured it was valid Python but didn't know how pytest really works.) Thanks.

>check separately that the automatic selection is working.

That may be tough. Certainly I can test for the failure of a solver by looking for the warning it produces then making sure the problem is successfully solved anyway. But this would require finding problems for which certain solvers consistently fail across platforms. I'll check the existing tests for a problem like this, but I can't generate one easily.

Reply:
Perhaps just check for each of the backends with a very strict tolerance. So if `has_sksparse` is True, then the automatic selection and forcing `sksparse` usage should give results that are numerically so close (because it's identical code) that you can check with something like `assert_allclose(..., rtol=1e-15)`. And such checks will fail with different backends normally.

Reply:
This will only test that the default solver is the one intended (it won't test whether `linprog` moves on to a different solver if the first one fails) but it's worth doing. Please see #10088.